src/main/java/com/example/petbuddybackend/service/rating/RatingService.java

```diff
@@ -0,0 +1,100 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class RatingService {
+
+    private final RatingRepository ratingRepository;
+    private final CareService careService;
+    private final UserService userService;
+    private final RatingMapper ratingMapper = RatingMapper.INSTANCE;
+
+    public Page<RatingResponse> getRatings(Pageable pageable, String caretakerEmail) {
+        return ratingRepository.findAllByCaretakerEmail(caretakerEmail, pageable)
+                .map(ratingMapper::mapToRatingResponse);
+    }
+
+    public RatingResponse rateCaretaker(String clientEmail, Long careId, int rating,
+                                        String comment) {
+
+        Care care = careService.getCareById(careId);
+        String caretakerEmail = care.getCaretaker().getEmail();
+        userService.assertCaretakerAndClientExist(caretakerEmail, clientEmail);
+        assertCareBetweenCaretakerAndClientIsValid(caretakerEmail, clientEmail, care);
+
+        if(caretakerEmail.equals(clientEmail)) {
+            throw new IllegalActionException("User cannot rate himself");
+        }
+
+        return ratingMapper.mapToRatingResponse(
+                createOrUpdateRating(caretakerEmail, clientEmail, careId, rating, comment)
+        );
+    }
+
+    public RatingResponse deleteRating(String clientEmail, Long careId) {
+        Care care = careService.getCareById(careId);
+        String caretakerEmail = care.getCaretaker().getEmail();
+        userService.assertCaretakerAndClientExist(caretakerEmail, clientEmail);
+
+        RatingKey ratingKey = new RatingKey(caretakerEmail, clientEmail, careId);
+        Rating rating = getRating(ratingKey);
+        ratingRepository.deleteById(ratingKey);
+
+        return ratingMapper.mapToRatingResponse(rating);
+    }
+
+    private Rating createOrUpdateRating(String caretakerEmail, String clientEmail, Long careId, int rating,
+                                        String comment) {
+        Rating ratingEntity = getOrCreateRating(caretakerEmail, clientEmail, careId);
+
+        ratingEntity.setRating(rating);
+        ratingEntity.setComment(comment);
+
+        return ratingRepository.save(ratingEntity);
+    }
+
+    private void assertCareBetweenCaretakerAndClientIsValid(String caretakerEmail, String clientEmail, Care care) {
+
+        if(!care.getClient().getEmail().equals(clientEmail) || !care.getCaretaker().getEmail().equals(caretakerEmail)) {
+            throw new IllegalActionException("This care does not exist between caretaker and client");
+        }
+
+        if(!care.getCaretakerStatus().equals(CareStatus.PAID) || !care.getClientStatus().equals(CareStatus.PAID)) {
+            throw new IllegalActionException("Cannot rate unpaid care");
+        }
+    }
```

Comment:
nie wiem czy `403` nie byłby lepszy dla obu wyjątków

Reply:
to jest chyba bardziej jak uzytkownik nie ma dostepu do zasobu, tutaj ma, tylko nie moze wykonac tej akcji

Reply:
Przy drugim wyjątku w sumie racja. Przy pierwszym to w sumie nie ma dostępu do opieki bo nie jest jego częścią.